\documentclass[nogrid]{MBE}%

\usepackage{url}
\usepackage{tabularx}
\usepackage{tablefootnote}
\jshort{mst}

\volname{}

\jvolume{0}

\jvol{}

\jissue{0}

\pubyear{2019}

\mstype{Letter}

\artid{HyPhy2.5}

\newcommand{\revised}[1]{{\color{red}#1}}
\newcommand{\sw}[1]{{\tt #1}}
\newcommand{\hyphy}{{\tt HyPhy}}
\newcommand{\hivtrace}{{\tt HIV-TRACE}}
\begin{document}

\title[HyPhy2.5]{HyPhy 2.5 -- a customizable platform for comparative sequence analysis using sophisticated evolutionary models.}

\author[Pond et al]{
	Sergei L.  \surname{Kosakovsky Pond}, Ryan Velazquez$^{1}$, Art F.Y. Poon,$^{2}$, Stephen Shank,$^{1}$,  Steven Weaver,$^{1}$,  N Lance Hepler, $^{3,4}$, Ben Murrell $^{3,5}$, Sadie Wisotsky,$^{1,6}$, Stephanie J Spielman, $^{1,7}$ Simon D.W. Frost, $^{8}$, Spencer V Muse $^{6}$
}

\address{$^{1}$Institute for Genomics and Evolutionary Medicine, Temple University, Philadelphia, PA, USA,
\\ $^{2}$Pathology and Laboratory Medicine - Western University, London, Ontario, Canada,
\\ $^{3}$University of California, San Diego, CA, USA
\\ $^{4}$10x Genomics...
\\ $^{5}$ Karolinska Institute, Stockholm, Sweden
\\ $^{6}$ North Carolina State University, Raleigh, NC, USA
\\ $^{7}$ Rowan University, Glassboro, NJ, USA
\\ $^{8}$ Department of Veterinary Medicine, Cambridge University, Cambridge, United Kingdom
}

\coresp{E-mail: spond@temple.edu}

\editor{TBD}

\abstract{
\hyphy{} (HYpothesis testing using PHYlogenies) is a scriptable open-source
package for fitting a  broad range of evolutionary models to multiple sequence
alignments and hypothesis testing primarily in the maximum likelihood
statistical framework. It has become a popular choice for characterizing various
aspects of the evolutionary process: natural selection, evolutionary rates,
recombination, and epistasis. The 2.5 release includes a completely
re-engineered computational core and analysis library that introduce new classes
of evolutionary models and statistical tests, delivers dramatic performance and
stability enhancements, improves usability, streamlines end-to-end analysis
workflows, and makes it easier to develop custom analyses. 
}

\keyword{evolutionary analysis, natural selection, hypothesis testing, statistical inference, software engineering}

\maketitle

\section{{Introduction}\label{sec:Intro}}

\hyphy{} is an open source software package for comparative sequence analysis
using stochastic evolutionary models. It is written in C++ and implements a
domain specific scripting language (HBL, or \hyphy{} Batch Language) similar in
syntax and conventions to JavaScript. \hyphy{} follows the common design
paradigm (e.g. R\cite{team2013r}, RevBayes\cite{hohna2016revbayes}, BEAST
2.x\cite{bouckaert2014beast}) of coupling a compiled computational engine with
compute-intensive core and relevant data types, with a rich and extensible
library of scripts or modules which are used to implement models, analyses, and
high-level algorithms. \hyphy{} is distributed with a sizable library of HBL
modules and pre-written analyses.  Since its initial release in 2005
[@doi:10.1093/bioinformatics/bti079] HyPhy has become an integral tool for the
bioinformatics community with over $2,500$ peer-reviewed citations and
approximately $1,000$ jobs processed each week on the companion Datamonkey web
application [@doi:10.1093/bioinformatics/bti320;
@doi:10.1093/bioinformatics/btq429; @doi:10.1093/molbev/msx335]. Extensions and
improvements to \hyphy{} have been ongoing since its inception, with active
feedback between users and developers producing new features tailored to the
specific needs of the research community. Here we announce the release of the
newest version of \hyphy{} (version 2.5) and document how the software has been
packaged for easy use in a variety of settings, optimized for larger datasets,
redesigned to follow modern  best practices, extended to include a broader set
of evolutionary models and pre-packaged analyses. 

The source code, installation instructions, and documentation for \hyphy{} is
available at \url{github.com/veg/hyphy} and \url{hyphy.org}, and the
accompanying JavaScript result visualization module -- at
\url{github.com/veg/hyphy-vision} and \url{vision.hyphy.org}. In addition, a public instance of the
\hyphy-powered web-application implementing the most popular analyses is
available at \url{www.datamonkey.org} \cite{weaver2018datamonkey}. 


\section{{New Approaches}\label{sec:Approaches}}

\subsection{Part of the software ecosystem}

Software development practices and usage patterns have decisively shifted away
from monolithic packages towards modular and reusable components that must be
installed, versioned, and maintained in complex software ecosystems \cite{xx}.
Recognizing that users of \hyphy{} vary greatly in their technical proficiency,
from biologists unfamiliar with the command line to bioinformaticians who want
to incorporate \hyphy{} into their own software, there are multiple ways to
interact with the package \revised{Ryan to sketch out a figure}

\begin{enumerate}
\item CLI ; keywords and interactive [simple example]
\item Electron app
\item Web applications; Datamonkey, Galaxy, MEGA
\item Library [currently broken]
\end{enumerate}

\subsection{Optimized for Larger Datasets} 

\hyphy{} has been optimized extensively tuned to enable analysis of modern
large-scale molecular data including methods. Core numerical and optimization
routines take advantage of vector processing (SIMD), multiprocessing (OpenMP),
and distributed systems (OpenMPI) for fine and coarse-grain parallelization as
appropriate to a specific analysis. We found that the majority of popular
analyses implemented in \hyphy{} do not significantly benefit from GPU
acceleration \cite{stamatakis2019review}, hence we currently do not offer this capability.
Even when fully accelerated, phylogenetic likelihood calculation for complex
models on large datasets remains a bottleneck which cannot be overcome with more
code tuning. However, we were able to integrate recent algorithmic advances from
the fields of machine learning and natural language processing \cite{blei2003latent} to
limit the number of expensive likelihood calculations for the key application of
high-throughput screening of alignments thousands of sequences for evidence of
natural selection \cite{murrell2013fubar}. Finally, we provide high performance computing
infrastructure[@doi:10.1093/bioinformatics/bti320;
@doi:10.1093/bioinformatics/btq429; @doi:10.1093/molbev/msx335] free of charge
to the global community, helping to democratize access to scientific computing
resources [@doi:10.1016/j.ijinfomgt.2013.05.010].

\subsection{Redesigned to Follow Modern Best Practices in Scientific Software  }

The design of the original \hyphy{} implementation emphasized convenient writing
and execution of single HBL scripts.  This decision manifested in the easy to
use command line prompt which guided users interactively through selecting the
desired analysis and choosing the specific analysis parameters.  Also, the HBL
itself was designed to allow sophisticated models to be specified and fit with
concise scripts, facilitated by extensive use of a global namespace.  As the
common use of \hyphy{} has shifted over the last decade from one-off analyses
toward larger studies and use within pipelines and web-applications, \hyphy{}
has  been redesigned to address the requirements of these changing use cases.   

Usage as a typical command line tool (i.e. an executable name followed by key
word arguments) has been added alongside the interactive command line prompt.
This change, along with the ability to use relative paths to files, has made
using HyPhy in pipelines and batch analyses seamless. In addition to being
easier to use, the package is also now easier to install. HyPhy is now
installable with \sw{bioconda} [@doi:10.1038/s41592-018-0046-7] and other
package managers, like \sw{homebrew}, which has become the defacto package
manager for scientific software. Users no longer need to concern themselves with
dependencies, environments and build processes but can simply type \sw{conda
install hyphy}.

The syntax, semantics, and structure of the standard HBL libraries have also
been re-designed and improved, providing a more reliable package that is easier
to extend and behaves more similarly to familiar language environments like
JavaScript.  Examples include: controlled vocabulary terms for molecular
evolutionary modeling, namespaces for larger library management, basic
functional-style programming (e.g. \sw{map, filter} etc), JavaDoc inline
documentation, etc ....

We are also standardizing the way \hyphy{} analyses operate. A command-line
analysis is expected to collect required inputs and options from the user,
either via keyword command line arguments, or prompts, execute the analysis,
report progress and key results to the screen (or file) in MarkDown format,
which can also be rendered as formatted report documents, and a complete set of
results to a JSON file. For the most popular analyses, we have developed a
JavaScript interactive visualization library (\url{vision.hyphy.org}) that reads
the JSON files and renders analysis-appropriate views (tables, charts,
alignments, trees, etc) of the results. JSON files can be readily parsed and
processed in \sw{Python}, \sw{R}, or any number of tools for subsequent
analysis. 

\subsection{Software engineering }

Software quality control and rigorous testing are expected to be a part of any
modern development cycle; a recent study shows this to be an aspirational goal
in the field of molecular evolution \cite{darriba2018state}. For \hyphy{}2.5 we have
implemented extensive automated testing as a part of a continuous integration
(CI) pipeline which which both expedites development and helps ensure healthy
software is delivered.  Our suite includes unit tests for over 90\% of HBL
functions, method tests on all the core analyses, and likelihood testing
[@url:https://gitlab.com/testiphy/testiphy] which compares the likelihood values
calculated by \hyphy{} with values calculated by other popular
maximum-likelihood software packages and informs the developers if discrepancies
are identified. For the core, we use C++ development and testing toolkits to
perform additional quality controls such as static code analysis, testing for
memory leaks and other issues, and instrumented code profiling. 

\subsection{Analyses and models supported}

Although the \hyphy{} package provides for near limitless customization via
writing HBL scripts, users can run many common analyses without needing to
concern themselves with the HBL at all.  The \hyphy{}  package comes with
pre-written HBL scripts for easily performing some of the most commonly used
analyses.  These analyses can be executed in the various places HyPhy is
available and include:  
  
\begin{itemize}
 \item
aBSREL[@doi:10.1093/molbev/msv022; @doi:10.1093/molbev/msr125] (adaptive Branch-Site Random Effects Likelihood) - Detect positive/diversifying selection at individual branches  
 \item 
BGM [@doi:10.1093/bioinformatics/btn313] (Bayesian Graphical Models) - Test for co-evolving sites  
 \item
 BUSTED [@doi:10.1093/molbev/msv035] (Branch-Site Unrestricted Statistical Test for Episodic Diversification) - Test gene-wide selection at pre-defined lineages  
 \item
 FADE (FUBAR Approach to Directional Evolution) - Evaluate if sites are subject to directional selection (not yet published) 
 \item
 FEL [@doi:10.1093/molbev/msi105] (Fixed Effects Likelihood) - Inferring non-synoymous and synonymous substitution rates on a per-site basis for smaller datasets using maximum likelihood  
 \item
 FUBAR [@doi:10.1093/molbev/mst030](Fast, Unconstrained Bayesian AppRoximation) - Infer non-synonymous and synonymous substitution rates on a per-site basis for larger datasets using Bayesian methods  
 \item
 GARD [@doi:10.1093/molbev/msl051] (Genetic Algorithm for Recombination Detection) - Screen multiple sequence alignment for recombination  
 \item
 MEME [@doi:10.1371/journal.pgen.1002764] (Mixed Effects Model of Evolution) - Test the hypothesis that individual sites have been subject to episodic positive/diversifying selection  
 \item
 RELAX  [@doi:10.1093/molbev/msu400] - Evaluate whether the strength of natural selection has been relaxed or intensified along a specific set of branches  
 \item
 SLAC [@doi:10.1093/molbev/msi105] (SingleLikelihood Ancestor Counting) - Infer non-synonymous and synonymous substitution rates on a per-site basis 
\end{itemize}


Add a description of different types of models and inference techniques
available. \revised{Sergei will write}


\section{Discussion}

Make \hyphy{} great again... COVFEFE!

\section{Acknowledgments}

This work was supported in part by grants R01 AI134384 (NIH/NIAID), R01 GM093939
(NIH/NIGMS) and U01 GM110749 (NIH/NIGMS). JOW was funded by an NIH-NIAID Career
Development Award (K01AI110181) and the California HIV/AIDS Research Program
(ID15-SD-052). 

\bibliographystyle{natbib}%%%%natbib.sty
\bibliography{hyphy2.5}%%%refs.bib

\end{document}
