## Extended to Include a Broader Set of Standard Analyses

Although the HyPhy package provides for limitless customization via writing HBL scripts, users can run many common analyses without needing to concern themselves with the HBL at all. 
The HyPhy package comes with pre-written HBL scripts for easily performing some of the most commonly used analyses. 
These analyses can be executed in the various places HyPhy is available and include (in alphabetical order):  
  
+ aBSREL [@doi:10.1093/molbev/msv022; @doi:10.1093/molbev/msr125] (adaptive Branch-Site Random Effects Likelihood) - Detecting positive selection at individual branches  
+ BGM [@doi:10.1093/bioinformatics/btn313] (Bayesian Graphical Models) - Testing for co-evolving sites  
+ BUSTED [@doi:10.1093/molbev/msv035] (Branch-Site Unrestricted Statistical Test for Episodic Diversification) - Testing gene-wide selection at pre-defined lineages  
+ FADE (FUBAR Approach to Directional Evolution) - Evaluating if sites are subject to directional selection  
+ FEL [@doi:10.1093/molbev/msi105] (Fixed Effects Likelihood) - Inferring non-synoymous and synonymous substitution rates on a per-site basis for smaller datasets using maximum likelihood  
+ FUBAR (Fast, Unconstrained Bayesian AppRoximation) - Infer non-synonymous and synonymous substitution rates on a per-site basis for larger datasets using Bayesian methods  
+ GARD [@doi:10.1093/molbev/msl051] (Genetic Algorithm for Recombination Detection) - Screen multiple sequence alignment for recombination  
+ MEME [@doi:10.1371/journal.pgen.1002764] (Mixed Effects Model of Evolution) - Test the hypothesis that individual sites have been subject to episodic positive/diversifying selection  
+ RELAX [@doi:10.1093/molbev/msu400] - Evaluate whether the strength of natural selection has been relaxed or intensified along a specific set of branches  
+ SLAC [@doi:10.1093/molbev/msi105] (SingleLikelihood Ancestor Counting) - Infer non-synonymous and synonymous substitution rates on a per-site basis 